datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,336 @@ fn optimize_join(
+                for input in &union.inputs {
+                    let mut replace_map = HashMap::new();
+                    for (i, field) in input.schema().fields().iter().enumerate() {
+                        replace_map.insert(
+                            union.schema.fields().get(i).unwrap().qualified_name(),
+                            Expr::Column(field.qualified_column()),
+                        );
+                    }
 
-            optimize_join(state, plan, left, right, on_filters)
-        }
-        LogicalPlan::TableScan(TableScan {
-            source,
-            projected_schema,
-            filters,
-            projection,
-            table_name,
-            fetch,
-        }) => {
-            let mut used_columns = HashSet::new();
-            let mut new_filters = filters.clone();
-
-            for (filter_expr, cols) in &state.filters {
-                let (preserve_filter_node, add_to_provider) =
-                    match source.supports_filter_pushdown(filter_expr)? {
-                        TableProviderFilterPushDown::Unsupported => (true, false),
-                        TableProviderFilterPushDown::Inexact => (true, true),
-                        TableProviderFilterPushDown::Exact => (false, true),
-                    };
-
-                if preserve_filter_node {
-                    used_columns.extend(cols.clone());
+                    let push_predicate =
+                        replace_cols_by_name(filter.predicate().clone(), &replace_map)?;
+                    inputs.push(Arc::new(LogicalPlan::Filter(Filter::try_new(
+                        push_predicate,
+                        input.clone(),
+                    )?)))
                 }
-
-                if add_to_provider {
-                    // Don't add expression again if it's already present in
-                    // pushed down filters.
-                    if new_filters.contains(filter_expr) {
-                        continue;
+                LogicalPlan::Union(Union {
+                    inputs,
+                    schema: plan.schema().clone(),
+                })
+            }
+            LogicalPlan::Aggregate(agg) => {
+                // An aggregate's aggregate columns are _not_ filter-commutable => collect these:
+                // * columns whose aggregation expression depends on
+                // * the aggregation columns themselves
+
+                // construct set of columns that `aggr_expr` depends on
+                let mut used_columns = HashSet::new();
+                exprlist_to_columns(&agg.aggr_expr, &mut used_columns)?;
+                let agg_columns = agg
+                    .aggr_expr
+                    .iter()
+                    .map(|x| Ok(Column::from_name(x.display_name()?)))
+                    .collect::<Result<HashSet<_>>>()?;
+                used_columns.extend(agg_columns);
+
+                let predicates = utils::split_conjunction_owned(utils::cnf_rewrite(
```

Comment:
I do not get the logic of Filter push down to `Aggregate` here. Why does it need to check the `agg_columns`?  Should we just check `group exprs` here ? And group exprs could introduce `Alias` (group by a + b as c), need to construct a replace_map from c - > a + b 
And if the filter expr's columns can be replaced by the replace map, they should be push predicates.

Reply:
Same with above, It's from original code😂.

I am also a little confused with these code. To avoid problems, I did not modify it.

You remind me😂.

As for `groupby`, I think original code forgot to handle it, it's a future ticket.

Reply:
You are re-implementing the rule, do not leave it for a future ticket.  The similar logic can be applied to Window partition by exprs.

Reply:
@mingmwang look like alias can't be in groupby.

sql 1999
```
<group by clause> Function
Specify a grouped table derived by the application of the <group by clause> to the result of the previously specified clause.
Format
<group by clause> ::=
       GROUP BY <grouping specification>
<grouping specification> ::=
         <grouping column reference>
       | <rollup list>
       | <cube list>
       | <grouping sets list>
       | <grand total>
       | <concatenated grouping>
```

pg:
```
-- create
CREATE TABLE EMPLOYEE (
  a INTEGER PRIMARY KEY,
  b INTEGER,
  c INTEGER
);

-- insert
INSERT INTO EMPLOYEE VALUES (0001, 0001, 0001);
INSERT INTO EMPLOYEE VALUES (0002, 0002, 0002);

-- fetch 
SELECT sum(b) FROM EMPLOYEE group by as c;

ERROR:  syntax error at or near "as"```